### ISSUE TITLE
YouTube bans IP addresses that use Streamlink

### ISSUE THREAD

**Issue by mtcdood** [streamlink/streamlink#2554]
## Plugin Issue
- [X] This is a plugin issue and I have read the contribution guidelines.

### Description
YouTube is now permanently blocking IP addresses that use Streamlink to download streams.

### Reproduction steps / Explicit stream URLs to test
1. Download a stream using Streamlink
2. Attempt to download another stream using Streamlink, can be hours/days later

### Log output
```
# streamlink --loglevel debug -o test.ts https://www.youtube.com/watch?v=YlJHBJNJu-0
[cli][info] streamlink is running as root! Be careful!
[cli][debug] OS:         Linux-4.9.0-8-amd64-x86_64-with-debian-9.8
[cli][debug] Python:     2.7.13
[cli][debug] Streamlink: 1.1.1
[cli][debug] Requests(2.22.0), Socks(1.7.0), Websocket(0.56.0)
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=YlJHBJNJu-0
[plugin.youtube][debug] Video ID from URL
[plugin.youtube][debug] Using video ID: YlJHBJNJu-0
error: Unable to open URL: https://youtube.com/get_video_info (429 Client Error: unknown for url: https://youtube.com/get_video_info?el=detailpage&video_id=YlJHBJNJu-0)
```

### Additional comments, screenshots, etc.
Just to establish this right fucking now, this is not my fault. I am not being rate limited because I downloaded too many streams or videos. I download less than one stream/video per day using Streamlink. Don't be like the massive asshole who develops youtube-dl whose codebase you share and permanently ban me from your Github repo for pointing out the issue is your code, not your users.

**Comment by beardypig** [comment]
Thank you for the issue report. You won't get banned for reporting issues, if keep the discourse civil everything will be hunky-dory. I don't believe we share a codebase with youtube-dl, I'm not sure where that idea came from - we might use a similar method for YouTube though, I personally haven't lifted anything from youtube-dl to use in streamlink, but I cannot say what other contributors might have done. Several people have contributed to the YouTube plugin, and may have drawn influence from the youtube-dl implementation. 

As I'm sure you can imagine, this will be a complex issue to debug. Clearly, Google/Youtube has a lot of resources to invest counter acting this type of software (streamlink, youtube-dl, etc.) and it is worth their while, because YouTube makes most of their money from ads. I know this set of people are probably the ones also using adblockers, but if people aren't visiting the website you cannot even up-sell them to YouTube Music of YouTube TV - which is maybe why it has started happening now, the services business is growing with Apple getting in the ring as well. 

YouTube _could_ be analysing the requests made from a single IP address to determine if they think it's a bot and ban the IP. For example, a normal user (using a web browser) would make a bunch of requests; images, extra javascript files, all sorts of things that Streamlink (and, I assume, youtube-dl) will not do. If you are using a server to run streamlink/youtube-dl this will be even more obvious to YouTube, as you won't have the regular browser requests. They could even have trained their fraud (we'll call it fraud, because I'm sure they do, and I would if I was EVP of Sales) detection algorithms to detect youtube-dl and streamlink - it wouldn't be hard, the request patterns are consistent and fair obvious. 

Is there a solution to this? I don't know... If YouTube put in effort to combat downloader tools, especially open source ones, I don't think there is much hope...

**Comment by scottbernstein** [comment]
This should be a relatively easy thing for us to prevent.  So long as we are passing a string which exactly matches a current web browser as the User-Agent, there is no way that YouTube can distinguish calls from Streamlink from a web browser.

**Comment by beardypig** [comment]
@scottbernstein, while you are correct in your assertion, it only applies for a single request. With multiple requests (or lack there of), a pattern to identify a non-browser client could be determined.

**Comment by scottbernstein** [comment]
This is true @beardypig

**Comment by johnthecracker** [comment]
How much youtube-dl and streamlink affect google?  1.000.000.000$ - 1$
This is how much they block me for now.

example streams that work:

https://www.youtube.com/embed/live_stream?channel= ( your stream )

https://www.youtube.com/embed/live_stream?channel=UC_yXEx3-Gk5drg9qhPzSQSg
https://www.youtube.com/embed/live_stream?channel=UCknLrEdhRCp1aegoMqRaCZg
https://www.youtube.com/channel/UCVgO39Bk5sMo66-6o6Spn6Q

and a few more.

```
$url = "https://www.youtube.com/embed/live_stream?channel=UC_yXEx3-Gk5drg9qhPzSQSg"
$iPID = Run('curl.exe ' & $url,'',@SW_HIDE,$STDOUT_CHILD)
ProcessWaitClose($iPID)
$s = StdoutRead($iPID)
$s = StringTrimLeft($s,StringInStr($s,'href="https://www.youtube.com/watch?v=')+5)
$s = StringTrimRight($s,(StringLen($s)+1)-StringInStr($s,'"'))
run('streamlink.exe --http-no-ssl-verify --player=mpv.exe ' & $s & ' 480p','',@SW_MINIMIZE)
```

The only benefit of streamlink here is that it handles streams better than players on their own.

No need for streamlink as primary tool to watch yoututbe streams.

Google would waste time, effort, money and brains to block streamlink.

**Comment by johnthecracker** [comment]
Your stream:

https://www.youtube.com/watch?v=YlJHBJNJu-0

```
C:\>streamlink.exe --player=mpv.exe --http-no-ssl-verify "https://www.youtube.com/watch?v=YlJHBJNJu-0" best
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/watch?v=YlJHBJNJu-0
error: This plugin does not support protected videos, try youtube-dl instead
```

I can watch this directly with mpv.

**Comment by mtcdood** [comment]
I never understood why the YouTube plugin claims it doesn't support protected videos. That should be removed. Every single time I've seen it it has been a non-protected video.

**Comment by beardypig** [comment]
@johnthecracker, this is all speculation but, they are likely working to detect fraud by looking at unusual usage behaviour, not specifically aimed at normal streamlink/youtube-dl users. I would say it’s more likely that you’ll get blocked if you use a VPS to download from YouTube and not use a browser on the VPS (or via VPN/proxy to the VPS), because you won’t have the normal user behaviour mixed it. YouTube are pretty famous for the false positives in their algorithms too... just look at the content id stuff :) That’s my guess. So my suggestion, just use streamlink/youtube-dl on the same connection as your browser (which is what I suspect the majority of users already do). I have not experienced any problems streaming YouTube using streamlink on my desktop. 

From what I understand, `mpv` uses `youtube-dl` under the hood.

**Comment by johnthecracker** [comment]
I use youtube-dl to download video clips and streamlink to watch live streams from youtube and
all that ever without web browser.
So far, I had no issues as an ordinary user practising that now and than.

To test and discuss defense tactics discussed here an ideal site would be one webcam site using finferprinting but webcam sites are no-no here.

Yes, mpv gets yotube streams via youtube-dl.

**Comment by beardypig** [comment]
@johnthecracker you never open YouTube.com?

**Comment by johnthecracker** [comment]
yes to get url for clips and no for stream channels as their url is constant.

watch?v= may change within channel url that's why I run curl to get right watch?v= one.

No browser use to watch live streams like Balkan News , DW, TRT, Sky and like.

**Comment by johnthecracker** [comment]
You may watch live streams that are on youtube dirctly on websites like AlJazeera, DW, TRT and many more. Only few ones like Balkan News are extra on youtube live.

If youtube wants to block Balkan guys from watching AlJazeera news, they may go for it.

**Comment by beardypig** [comment]
In my version of this imaginary system you would be classified as a normal user; you make requests for images, JavaScript files, css, etc. all the things a regular user does. Your use of the YouTube website doesn’t particularly out of the ordinary, thus no ban. My point is, if you go to the website, grab the link and stream it on the same pc/internet connection you’ll likely be fine. If you copy the link, then download the video on a remote server, you might get banned (on the remote server) because you don’t have normal user behaviour.

@mtcdood how do _you_ use YouTube and streamlink together?

**Comment by mtcdood** [comment]
I simply run Streamlink on a remote server when I'm informed a live stream I want to download is running. I don't make any attempts to obfuscate that it's automated activity. I used this approach from mid-March through July 10 before finding the server blocked. I've successfully downloaded one stream since then. All others (one every few days) have failed.

**Comment by bastimeyer** [comment]
Since @mtcdood is running Streamlink as root, it looks like he's not using it on a regular user system (desktop/laptop/etc) but on a rented server instead (debian amd64) and it's possible that the entire address-block and/or hostnames related to it got blocked. This could have been caused either by the abuse of him and his server(s), or by somebody else on that same network.

However, Streamlink and the Youtube plugin alone do not result in a ban from Youtube. If that was the case, then we would have seen many more of these claims here on the issue tracker over the past years.

I also don't think that YouTube is evaluating HTTP requests of the various assets on their website (JS/CSS/fonts/images). Since these CDN requests are being cached by the browser, it makes it hard to believe that missing requests when using Streamlink could be the cause of a ban like this.

**Comment by mtcdood** [comment]
The fact remains that JDownloader is not blocked, Streamlink is.

**Comment by bastimeyer** [comment]
What makes Youtube able to indentify you as a Streamlink user (or as a bot in that same sense) is that Streamlink's Youtube plugin is using an *old*, static User-Agent HTTP header:
https://github.com/streamlink/streamlink/blob/1.1.1/src/streamlink/plugins/youtube.py#L175
https://github.com/streamlink/streamlink/blob/1.1.1/src/streamlink/plugin/api/useragents.py#L3

You can try overriding this header by using the `--http-header` parameter.
https://streamlink.github.io/cli.html#cmdoption-http-header
But since you've already been blocked, I'm not sure if this will have an effect.

**Comment by beardypig** [comment]
@mtcdood, is JDownloader something you run on your local machine? 

@bastimeyer, they would have CDN logs (and click stream data) - it's not outside the realms of possibility that they are doing something similar to what I have suggested. They probably wouldn't rely on a user agent string along, but it could be a component in an algorithm. They are doing a lot of analysis on the visitors to the website, they have engineering teams dedicated to producing stats for analysts to look at, with teams looking at the reports the analysts generate... It could also be abuse of others on the same network.

**Comment by mtcdood** [comment]
@beardypig I run JDownloader on the same VPS that I run Streamlink on.

**Comment by beardypig** [comment]
I just checked out what JDownloader does, and it behaves slightly differently to Streamlink. Streamlink only requests the `get_video_info` endpoint, where as JDownloader loads the video page, grabs an image, grabs the subtitles, then it gets `get_video_info`. `youtube-dl` requests the video page, and the multiple versions of `get_video_info`. 🤷‍♂

**Comment by thinkpad4** [comment]
A user agent of Chrome 71 is old?  I use a Chrome Browser version 47 to watch youtube and I don't have any issues.  There must be more going on than just them reading the user agent

**Comment by bastimeyer** [comment]
Yes, Chrome 71 is old in the sense that it's not the current stable version, which is 75. The Chromium release cycle is approximately 6 weeks, and 71 has been released on Dec 4th 2018. The higher the difference between the stable version gets, the more specific HTTP requests using this User-Agent header get. This means that Youtube able to identify tools like Streamlink, since 99.9% of all Chrome users will follow Google's release cycle.

**Comment by scottbernstein** [comment]
Also FYI, I do my share of pointing streamlink at YouTube, and I have never seen evidence of being blocked...

**Comment by Viat0r** [comment]
I have the same problem.
Perhaps Youtube blocks IP addresses it considers 'suspicious', but it's not streamlink's fault because youtube-dl is not working either.
I've just used streamlink with a proxy and it worked fine!

I recorded quite a few live YT streams recently, so this might be the reason my IP got blocked.

**Comment by Viat0r** [comment]
Dream functionality: an option to imitate the real browser behavior!

**Comment by johnthecracker** [comment]
If you just add a cookie to your call that might just do..

**Comment by liamkennedy** [comment]
I'm not seeing any banning of youtube (yet) and I stream a lot (NASA HDEV via their Youtube channel).  

I hope it stays that way.  I'll definitely be monitoring this issue to check on the verification of this.

**Comment by gravyboat** [comment]
@liamkennedy I don't think you're using a headless server in some DC though right? You're still just using the Pi boxes that people have on a network that likely has real Youtube traffic as well.

**Comment by liamkennedy** [comment]
@gravyboat yes - that's exactly the usage model - so I understand the major difference in use case here.  Thanks for clarifying that aspect of the issue.

**Comment by limiteddi** [comment]
It looks like YouTube gives 429 when live stream ends and streamlink keeps banging the server for a video, if you use some for of monit.

**Comment by beardypig** [comment]
@limiteddi using which options?

**Comment by limiteddi** [comment]
> @limiteddi using which options?

No options for links, just http output.

**Comment by beardypig** [comment]
That’s an option ;) Can you post the full command you’re using, so I can try to reproduce the ban.

**Comment by limiteddi** [comment]
`/usr/local/bin/streamlink "https://www.youtube.com/channel/UCNvCQpcafnbW4KQ8X7oQ9kg" best --player-external-http --player-external-http-port 12345`

It's gotten ridiculously difficult to stream something because of their new measures.

**Comment by beardypig** [comment]
Which player are you using @limiteddi ?

**Comment by johnthecracker** [comment]
<a href='https://postimg.cc/SnwJ8dqj' target='_blank'><img src='https://i.postimg.cc/SnwJ8dqj/Screen-Shot.jpg' border='0' alt='Screen-Shot'/></a>

**Comment by johnthecracker** [comment]
streamlink-1.1.1-20190724.exe

youtube-dl 2019.07.30

```
C:\>streamlink.exe https://www.youtube.com/channel/UCNvCQpcafnbW4KQ8X7oQ9kg best
[cli][info] Found matching plugin youtube for URL https://www.youtube.com/channel/UCNvCQpcafnbW4KQ8X7oQ9kg
[cli][info] Available streams: 144p (worst), 240p, 360p, 480p (best)
[cli][info] Opening stream: 480p (hls)
[cli][info] Starting player: C:\P\VLC\vlc.exe
```

**Comment by beardypig** [comment]
@mtcdood has the force IPv4 option suggested in various youtube-dl forums worked for you?

**Comment by mtcdood** [comment]
Yes.

**Comment by beardypig** [comment]
@mtcdood thanks for reporting back. Let's see if we can get a `--force-ipv{4,6}` option in to `streamlink`...

**Comment by limiteddi** [comment]
> @mtcdood thanks for reporting back. Let's see if we can get a `--force-ipv{4,6}` option in to `streamlink`...

Hello, would you have any ETA on this?

**Comment by ChrisLX200** [comment]
Would like to know if this is a viable solution to the current problem with YT.

**Comment by limiteddi** [comment]
I havent managed to find a viable solution. Some scripting checks for live feed if available, delays in retry etc help temporarily but eventually it throws out 429 :(

**Comment by TPME** [comment]
I have the same issue, none of the YouTube Livestreams work. It looks like an option to force IPV4 is the solution (works for `youtube-dl`).

Please help!

**Comment by gravyboat** [comment]
There's nothing more to discuss on this issue so I'm going to lock it. We'll see about adding an ipv4 option to Streamlink when someone feels like working on it.
